### ISSUE TITLE
Enable logging during bootstrap phase

### ISSUE THREAD

**Issue by abhinavdahiya** [kubernetes-incubator/bootkube#557]
Right now the only way to access the kubelet logs and logs of the static pods in the bootstrap control plane on the bootstrap node is to ssh into the node!

It would ideal if we could deploy something like a static fluentd pod on the bootstrap node to upload all the logs to something like a s3 bucket for debugging purposes.

/cc: @aaronlevy @yifan-gu @diegs

**Comment by abhinavdahiya** [comment]
fluentd is a good option for extracting logs as it have various configurable input & output options, esp output to s3 bucket.

fluentd needs a fluent.conf to configure sources & outputs

Some ways we can implement is
- static pod during bootstrap
** fluentd.conf is shipped in asset-dir
** bake fluentd.conf in a custom fluentd-image 
- non static pod after api-server is up.

**Comment by yifan-gu** [comment]
At the first glance, a custom fluentd image seems cleaner than shipping config file in the asset-dir given that it hides the details and the config file should be pretty stable. The cons are that if we do want to make a change on conf file or fluentd image, we need to bake a new image.

Non-static pod might be a less ideal option as it will depend on api server to be running before it can be deployed. However, "why api server is NOT running" is the exact problem we want to debug with the logging.

**Comment by diegs** [comment]
It seems like there are 2 main areas we want to address.

(1) Issues during bootstrap, e.g. the boostrap control plane doesn't come up.
(2) Issues with the self-hosted control plane and downstream operations (this is not bootkube-specific and is more relevant to tectonic-installer, various operators, and test frameworks).

For (1) I think a static fluentd pod that can hook up to something easily accessible on the bootstrap node (e.g. journal logs) would suffice and actually be preferable. Bootkube requires you to ssh to that node anyway to run it. I don't think it's within the scope of the bootkube project to add more dependencies on logging sinks, such as S3 buckets. Currently the bootkube e2e tests fetch the journal logs at the end of the test--that's a simple command to run which we could publish in docs.

For (2) I think we should just recommend that people to deploy a more permanent fluentd dameonset as part of the asset-dir/manifests directory. But again, I think this is environment-specific and doesn't belong in the bootkube project, except perhaps as an example in the hack-scripts.

So to summarize, I think doing something simple for (1) that doesn't add external infrastructure dependencies sounds good, and I think it would be sufficient to just make sure the pod logs are aggregated in a place that's easy to extract.

**Comment by pbx0** [comment]
Current logging code in the pluton e2e tests uses the package: [github.com/coreos/mantle/network/journal](https://github.com/coreos/mantle/tree/540ba76432795d30c15c71003c8c070c1f7d6876/network/journal) to stream the journal logs as the test runs and then picks back up at the last cursor during reboots. Might be overkill for bootkube needs, but it could be re-used in the current (non-pluton) framework. Just need to dump pod logs at the end as well. You can see use of the package [here](https://github.com/coreos/mantle/blob/540ba76432795d30c15c71003c8c070c1f7d6876/platform/journal.go)

Agreed that having external platform dependancies like s3 might make things a little complicated. I like the ergonomics of having logs dropped locally when running tests on dev machines.  

On the other hand our current ability to SSH into the external IPs of nodes is not a feature that will likely be kept as a more common framework emerges since many tectonic platforms will not have nodes with external IPs exposed. If we find a way to SSH to nodes through the API via special pod that tunnels out SSH access then we could rely on this transport for our logging needs.
